### PR TITLE
Add contract calls to get config options

### PR DESCRIPTION
**Pull request by jferrant**
Depends on https://github.com/Trust-Machines/core-eng/issues/325
Closes https://github.com/Trust-Machines/core-eng/issues/216
Closes https://github.com/Trust-Machines/core-eng/issues/300

Useful notes for testing coordinator:

1. Ensure GCloud is configured by following the steps in this README: https://github.com/Trust-Machines/core-eng-private#readme
2. Useful GCloud commands:
- Shell into the VM: 
`gcloud compute ssh --zone "us-central1-a" "user@sbtc" --project "core-eng"`
- Port forward stacks node (note it is running next-master-merge which contains alpha breaking changes):
`gcloud compute ssh --zone "us-central1-a" "user@sbtc" --project "core-eng" -- -NL 20443:localhost:20443`
- Ensure your config file points to this node:
`stacks_node_rpc_url = "http://localhost:20443"`
3. Set up a bitcoin node for testing. Ensure your config file points to this node
- `bitcoin_node_rpc_url = "http://abcd:abcd@localhost:18445"`
- `bitcoind -rpcuser=abcd -rpcpassword=abcd -rpcport=18445`
4. Example peg in and out requests: https://testnet.stx.eco/history?net=testnet
- http://localhost:20443/v2/burn_ops/2425663/peg_out_request
- http://localhost:20443/v2/burn_ops/2425540/peg_in
5. sbtc alpha is deployed to testnet. Feel free to explore it at: https://explorer.hiro.so/sandbox/contract-call/ST306HDPY54T81RZ7A9NGA2F03B8NRGW6Y59ZRZSD.faint-tan-cobra/get-signer-data?chain=testnet
6. Set up signers in background:
cargo run --bin relay-server
`stacks-signer run --id 1 --config signer.toml`
`stacks-signer run --id 2 --config signer.toml`
`stacks-signer run --id 3 --config signer.toml`
7. Run coordinator
`stacks-coordinator --config ~/git/core-eng/stacks-coordinator/conf/coordinator.toml --signer-config ~/git/core-eng/stacks-coordinator/conf/signer.toml -b 2425540 run`
NOTE: the `-b` option is to specify a starting block height which is useful for forcing it to quickly find a block height that has a peg in or peg out...
NOTE: if you don't change the target_block_height within the code when it polls for peg-ins and outs...you are going to wait a loooong time as it iterates from start block height to the current block height which at this point if you use the known sbtc transaction block heights....is quite far behind the tip.

### CODE REVIEW

File: stacks-coordinator/src/stacks_wallet.rs [hunk 1]
```diff
@@ -278,13 +246,31 @@ mod tests {
...
+        let sender_key =
+            StacksPrivateKey::from_hex(PRIVATE_KEY_HEX).expect("Failed to parse private key");
+
+        let pk = StacksPublicKey::from_private(&sender_key);
+
+        let address = StacksAddress::from_public_keys(
+            address_version(&TransactionVersion::Testnet),
+            &AddressHashMode::SerializeP2PKH,
+            1,
+            &vec![pk],
```
Comment (stjepangolemac) [inline stacks-coordinator/src/stacks_wallet.rs:258]: It might not need `vec!` here if it expects a slice.

**Review by stjepangolemac — COMMENTED** [review]
Didn't find anything suspicious on first glance. Will make sure to review once more when it's ready.

**Review by netrome — COMMENTED** [review]
Some nits and style thoughts along the way. Let me know when the final PR is ready and I'll re-review.

File: stacks-coordinator/src/stacks_wallet.rs (hunk 1 above)
Comment (jferrant) [reply stacks-coordinator/src/stacks_wallet.rs:258]: It requires a &Vec<T> :/

File: stacks-coordinator/conf/coordinator.toml [hunk 2]
```diff
@@ -1,8 +1,9 @@
-sbtc_contract = "SP2BJA4JYFJ7SDMNFJZ9TJ3GB80P9Z80ADPGK1C2F.sbtc-alpha"
+sbtc_contract = "ST3N4AJFZZYC4BK99H53XP8KDGXFGQ2PRSPNET8TN.sky-blue-elephant"
```
Comment (xoloki) [inline stacks-coordinator/conf/coordinator.toml:1]: Does this point to your recent contract PR's deployment?  If so, can you change it back now that's it's been merged?

**Review by xoloki — CHANGES_REQUESTED** [review]
One question for now, will review more thoroughly after the PR is out of draft.

File: stacks-coordinator/src/stacks_node/client.rs [hunk 3]
```diff
@@ -132,6 +287,137 @@ impl StacksNode for NodeClient {
...
+            contract_address,
+            contract_name,
+            function_name,
+            vec![],
+        )?;
+        let coordinator_data = ClarityValue::try_deserialize_hex_untyped(&coordinator_data_hex)?;
+        if let ClarityValue::Optional(optional_data) = coordinator_data.clone() {
+            if let Some(data) = optional_data.data {
+                if let ClarityValue::Tuple(tuple_data) = *data {
+                    let value = tuple_data
```
[the remark this answers is not in the document]
Comment (netrome) [reply stacks-coordinator/src/stacks_node/client.rs:323]: Ah I missed that. But you can include the struct in the match statement. Something like:
```
if let ClarityValue::Optional(OptionalData{data: Some(ClarityValue::Tuple(tuple_data)), ..}) = ...
```

[the remark this answers is not in the document]
Comment (netrome) [reply stacks-coordinator/src/stacks_node/client.rs]: Yeah that seems to be unstable https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html

File: stacks-coordinator/src/stacks_node/client.rs [hunk 4]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+        );
+        let result = h.join().unwrap().unwrap();
+        assert_eq!(result, "0x070d0000000473425443");
+    }
+
+    #[test]
+    fn call_read_failure_test() {
+        let config = TestConfig::new();
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:465]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 5]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+        let result = h.join().unwrap();
+        assert!(matches!(result, Err(StacksNodeError::ReadOnlyFailure(_))));
+    }
+
+    #[test]
+    fn signer_data_none_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:483]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 6]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+        let result = h.join().unwrap();
+        assert!(matches!(result, Err(StacksNodeError::NoSignerData(_))));
+    }
+
+    #[test]
+    fn keys_threshold_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:503]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 7]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+        let result = h.join().unwrap().unwrap();
+        assert_eq!(result, 2800);
+    }
+
+    #[test]
+    fn keys_threshold_invalid_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:517]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 8]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+            Err(StacksNodeError::MalformedClarityValue(..))
+        ));
+    }
+
+    #[test]
+    fn num_signers_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:536]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 9]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+        let result = h.join().unwrap().unwrap();
+        assert_eq!(result, 4000);
+    }
+
+    #[test]
+    fn num_signers_invalid_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:551]:
```suggestion
```

File: stacks-coordinator/src/stacks_node/client.rs [hunk 10]
```diff
@@ -144,18 +376,258 @@ mod tests {
...
+            Err(StacksNodeError::MalformedClarityValue(..))
+        ));
+    }
+
+    #[test]
+    fn next_nonce_success_test() {
+        let config = TestConfig::new();
+
+        let h = spawn(move || {
+            sleep(Duration::from_millis(100));
```
Comment (netrome) [inline stacks-coordinator/src/stacks_node/client.rs:570]:
```suggestion
```

**Review by netrome — APPROVED** [review]
LGTM. Just some redundant `thread::sleep` in tests calls I'd like to see removed.

**Comment by jferrant** [comment]
> 

Good catch. This was a "copy/paste without thinking" moment

**Review by stjepangolemac — APPROVED** [review]

**Pull request merged**